sys/kern/proc.c

```diff
@@ -569,42 +569,76 @@ __noreturn void proc_exit(int exitstatus) {
   thread_exit();
 }
 
+static int proc_pgsignal(cred_t *cred, session_t *s, pgid_t pgid, signo_t sig) {
+  pgrp_t *pgrp = NULL;
+
+  WITH_MTX_LOCK (all_proc_mtx) {
+    if (pgid == 0) {
+      pgrp = proc_self()->p_pgrp;
+    } else {
+      pgrp = pgrp_lookup(pgid);
+      if (!pgrp)
+        return ESRCH;
+    }
+    mtx_lock(&pgrp->pg_lock);
+  }
+
+  proc_t *target;
+  int send = 0, error = 0;
+  TAILQ_FOREACH (target, &pgrp->pg_members, p_pglist) {
+    WITH_PROC_LOCK(target) {
+      if (!(error = proc_cansignal(cred, s, target, sig))) {
+        sig_kill(target, sig);
+        send++;
+      }
+    }
+  }
+  mtx_unlock(&pgrp->pg_lock);
+
+  /* We return error when signal can't be send to any process. Returned error is
+   * last error obtained from checking privileges.*/
+  return send > 0 ? 0 : error;
+}
+
 int proc_sendsig(pid_t pid, signo_t sig) {
 
+  if (sig >= NSIG)
+    return EINVAL;
+
+  int error;
+  proc_t *p = proc_self();
   proc_t *target;
+  session_t *session;
+
+  WITH_PROC_LOCK(p) {
+    /* Take a snapshot of session ptr. We will use it in privilege check. */
+    session = p->p_pgrp->pg_session;
+  }
```

Comment:
Not needed if we call `proc_cansignal()` with `all_proc_mtx` held, which I suggest we do.